src/realm/object-store/sync/sync_manager.cpp

```diff
@@ -521,13 +521,19 @@ SyncManager::~SyncManager()
     }
 #endif
 
-    for (auto& user : m_users) {
-        user->detach_from_sync_manager();
+    {
```

Comment:
@jbreams I believe that this was an oversight. Every access to **m_users** and **m_sync_client** is protected by a mutex.

Reply:
I think the assumption here is that if the SyncManager is being destructed, then the last shared pointer to this instance has just been released and there can be no other accessors. But guarding with the mutex here couldn't hurt.

Reply:
I agree, but I think the same can be argued about **m_sessions** above.

Reply:
I think by adding an anchor to the login functions in app.cpp this will probably never actually race. The logic for m_sessions was more complicated because while the sync manager and app are being destroyed sync sessions are also getting closed/destroyed and so the list of sessions can change. I think if we're in this destructor then the app is being destroyed too which means there are no outstanding callbacks that would login/logout a user.

All that said, adding a lock here doesn't seem harmful.